src/server/worker/master.go

```diff
@@ -531,366 +526,241 @@ func plusDuration(x *types.Duration, y *types.Duration) (*types.Duration, error)
 	return types.DurationProto(xd + yd), nil
 }
 
-// jobManager feeds datums to jobs
-func (a *APIServer) runJob(ctx context.Context, jobInfo *pps.JobInfo, pool *pool.Pool, logger *taggedLogger) error {
-	pfsClient := a.pachClient.PfsAPIClient
-	ppsClient := a.pachClient.PpsAPIClient
+func (a *APIServer) locks(jobID string) col.Collection {
+	return col.NewCollection(a.etcdClient, path.Join(a.etcdPrefix, lockPrefix, jobID), nil, &types.Empty{}, nil)
+}
 
-	jobID := jobInfo.Job.ID
-	var jobStopped bool
-	var jobStoppedMutex sync.Mutex
-	backoff.RetryNotify(func() (retErr error) {
-		// We use a new context for this particular instance of the retry
-		// loop, to ensure that all resources are released properly when
-		// this job retries.
-		ctx, cancel := context.WithCancel(ctx)
-		defer cancel()
+// collectDatum collects the output and stats output from a datum, and merges
+// it into the passed trees. It errors if it can't find the tree object for
+// this datum, unless failed is true in which case it tolerates missing trees.
+func (a *APIServer) collectDatum(ctx context.Context, index int, files []*Input, logger *taggedLogger,
```

Comment:
I think breaking up runJob will make this much easier to read

Reply:
I took a crack at break this up and found the improvements weren't that great. The code got more organized looking but the downside was that the functions I broke it up into wound up needing to take like 10 arguments each just to have the variables they needed present.